### ISSUE TITLE
[feature request] `source $(nixGL shell)` support

### ISSUE THREAD

**Issue by samuela** [nix-community/nixGL#156]
It would be super handy to be able to "install" nixGL into a shell session such that I don't have to worry about wrapping each command in `nixGL foo ...`.

Would maintainers be supportive of including such a feature?

**Comment by jim3692** [comment]
You can alias `nix-shell` to `nixGL nixVulkanIntel nix-shell`

**Comment by samuela** [comment]
That's certainly an option, but what I'm ideally looking for is a solution to "install" nixGL in the currently running shell instead of spawning a new one.

**Comment by jim3692** [comment]
> That's certainly an option, but what I'm ideally looking for is a solution to "install" nixGL in the currently running shell instead of spawning a new one.

Lately Arch updated its gcc version and now that alias fails with these errors:
```
nix-shell: /nix/store/bn7pnigb0f8874m6riiw6dngsmdyic1g-gcc-13.3.0-lib/lib/libstdc++.so.6: version `CXXABI_1.3.15' not found (required by /usr/lib/libnixexpr.so)
nix-shell: /nix/store/bn7pnigb0f8874m6riiw6dngsmdyic1g-gcc-13.3.0-lib/lib/libstdc++.so.6: version `CXXABI_1.3.15' not found (required by /usr/lib/libnixmain.so)
nix-shell: /nix/store/bn7pnigb0f8874m6riiw6dngsmdyic1g-gcc-13.3.0-lib/lib/libstdc++.so.6: version `CXXABI_1.3.15' not found (required by /usr/lib/libnixstore.so)
nix-shell: /nix/store/bn7pnigb0f8874m6riiw6dngsmdyic1g-gcc-13.3.0-lib/lib/libstdc++.so.6: version `CXXABI_1.3.15' not found (required by /usr/lib/libnixutil.so)
```

So, I came up with a more "install" solution:
```bash
if [[ ! "$IN_NIX_SHELL" = "" ]] && [[ "$IN_NIXGL" = "" ]]; then
  export IN_NIXGL=1
  source <(
    diff <(env) <(nixGL nixVulkanIntel env) \
      | grep '>' | sed 's/^>/export/g'
  )
fi
```

It extracts the environment variables that `nixGL` and `nixVulkanIntel` set, and adds them to the current shell.

**Comment by theVakhovskeIsTaken** [comment]
Was looking at the solutions posted before and after, and in cross-repositories, but decided to make it the "NixOS" way, so that I won't boil my mind

Since I'm running an immutable setup, I wanted to keep the "packaging" unhingeness for personal reasons

The thing: https://github.com/immutarch/postcopy_beta/blob/main/usr/bin/ima-nix-mesa-binder

What this does?
1. Runs `nixGL env`, checks for `LIBGL_LIBRARY_PATH` entries, where usually both 64-bit and 32-but MESA outputs are located
2. Determines which arch are they, and appropiately symlinks each store to `/run/opengl-driver(-32)`

Cons:
1. You might have to manually re-add symlinks after each mesa + nixgl update

Other than that, it works very much fine with everything graphically-required installed

![IMG_20240725_022529_491](https://github.com/user-attachments/assets/b1fb9a1d-decb-4305-812f-ba938f037210)
![IMG_20240725_022534_403](https://github.com/user-attachments/assets/ad739419-4e41-4593-a44b-d34209ccd5f9)
![IMG_20240725_022610_475](https://github.com/user-attachments/assets/107862fb-f8a3-4ed9-99d5-445408a5a12c)

**Comment by jim3692** [comment]
> 2. Determines which arch are they, and appropiately symlinks each store to `/run/opengl-driver(-32)`

I can't understand how the games manage to find the OpenGL drivers in that path. Are you exporting the `/run/opengl-driver` path?

**Comment by theVakhovskeIsTaken** [comment]
> Are you exporting the /run/opengl-driver path?

No, I am not. The base library preload directory is defined by package on it's own, so this entire thing is almost 1:1 with NixOS in terms of how drivers get loaded

**Comment by jim3692** [comment]
Great. I will check it. Is there any related documentation to that? I want to understand whether something like this is possible for Vulkan.

**Comment by theVakhovskeIsTaken** [comment]
> Great. I will check it. Is there any related documentation to that? I want to understand whether something like this is possible for Vulkan.

No reason to, these directories already include Vulkan drivers and their ICD loaders

**Comment by jim3692** [comment]
Sorry. My fault. I expected Vulkan to require a different method, as it's exposed with `nixVulkanIntel`, and not `nixGL`. Your script seems to work fine on Arch.

**Comment by theVakhovskeIsTaken** [comment]
```bash
#!/bin/bash

mkdir -p /etc/tmpfiles.d
get_dri_info(){
    MESA_DRIVERS_PATH=$(nixGL env | grep LIBGL_DRIVERS_PATH | cut -d '=' -f 2 | sed 's/\/lib\/dri//g' | sed 's/:/ /g')
    for driver in ${MESA_DRIVERS_PATH}
    do
        DRIVER_FILE=$(file ${driver}/lib/libEGL_mesa.so.0.0.0 | cut -d ' ' -f 3)
        if [[ "${DRIVER_FILE}" == "64-bit" ]]; then
            DRIVER_64BIT=${driver}
        else
            DRIVER_32BIT=${driver}
        fi
    done
    echo "64-bit driver path: ${DRIVER_64BIT} | This will be symlinked to /run/opengl-driver"
    echo "L+ /run/opengl-driver - - - - ${DRIVER_64BIT}" > /etc/tmpfiles.d/99-nix-ogl-64b.conf
    echo "32-bit driver path: ${DRIVER_32BIT} | This will be symlinked to /run/opengl-driver-32"
    echo "L+ /run/opengl-driver-32 - - - - ${DRIVER_32BIT}" > /etc/tmpfiles.d/99-nix-ogl-32b.conf
    systemd-tmpfiles --create
}

if [[ "$(which nixGL)" ]]; then
    get_dri_info
else
    echo "nixGL is not installed on this system, please install nixGL"
fi
```

I've updated my scripts for my distro, so that's why i'm sending an universally working one here

**Comment by soupglasses** [comment]
There's now also https://github.com/soupglasses/nix-system-graphics which is an alternative approach to do this with, which utilizes the `/run/opengl-driver` approach instead of environment variable magic as nixGL does.

**Comment by jim3692** [comment]
I started using home-manager. I don't plan to move to system-manager at the moment, so I did the following:

- Created `/etc/tmpfiles.d/99-nix-ogl.conf` with the following contents. Make sure to replace the `{USERNAME}`.
  ```
  L+ /run/opengl-driver-32 - - - - /nix/var/nix/profiles/per-user/{USERNAME}/profile/drivers/opengl-driver-32
  L+ /run/opengl-driver - - - - /nix/var/nix/profiles/per-user/{USERNAME}/profile/drivers/opengl-driver
  ```
- Added nixgl to the flake inputs and extraSpecialArgs:
  ```nix
  {
    inputs = {
      # ...

      nixgl = {
        url = "github:nix-community/nixGL";
        inputs.nixpkgs.follows = "nixpkgs";
      };
    };

    outputs = { home-manager, nixgl, ... }: {
      homeConfigurations."{USERNAME}" = home-manager.lib.homeManagerConfiguration {
        # ...
        extraSpecialArgs = { inherit nixgl; };
      };
    };
  }
  ```

- Added the following configuration to home-manager:
  ```nix
  home.packages = [
    nixgl.packages.${pkgs.system}.nixGLIntel
    nixgl.packages.${pkgs.system}.nixVulkanIntel

    (
      let
        driversEnv = with pkgs; buildEnv {
          name = "graphics-drivers";
          paths = [ mesa.drivers rocmPackages.clr.icd ];
        };

        driversEnv32 = with pkgs; buildEnv {
          name = "graphics-drivers-32bit";
          paths = with pkgsi686Linux; [ mesa.drivers ];
        };
      in pkgs.runCommand "graphics-drivers" {} ''
        mkdir -p $out/drivers ; cd $out/drivers
        ln -s "${toString driversEnv32}" opengl-driver-32
        ln -s "${toString driversEnv}" opengl-driver
      ''
    )
  ]
  ```

Now, I can update the Mesa/ROCm drivers for Nix using home-manager, without requiring root privileges.

It's inspired by the way NixOS installs OpenGL/Vulkan/OpenCL drivers.
For more info:
[64-bit drivers](https://github.com/NixOS/nixpkgs/blob/689fed12a013f56d4c4d3f612489634267d86529/nixos/modules/hardware/opengl.nix#L13)
[32-bit drivers](https://github.com/NixOS/nixpkgs/blob/689fed12a013f56d4c4d3f612489634267d86529/nixos/modules/hardware/opengl.nix#L18)
[tmpfiles](https://github.com/NixOS/nixpkgs/blob/689fed12a013f56d4c4d3f612489634267d86529/nixos/modules/hardware/opengl.nix#L141)

**Comment by juipeltje** [comment]
@jim3692 thanks for sharing that. this opengl problem has been bugging me for the longest time, and i couldn't use system-manager because it relies on systemd, and i like using void linux, which doesn't use systemd. your solution seems to work perfectly. i can't symlink with tmpfiles.d but it should be no problem to just set up a cron job on void or something like that instead. just wanted to ask though: with this solution you don't need to use nixgl at all anymore right? since you mentioned it as part of the steps.

**Comment by jim3692** [comment]
> @jim3692 thanks for sharing that. this opengl problem has been bugging me for the longest time, and i couldn't use system-manager because it relies on systemd, and i like using void linux, which doesn't use systemd. your solution seems to work perfectly. i can't symlink with tmpfiles.d but it should be no problem to just set up a cron job on void or something like that instead. just wanted to ask though: with this solution you don't need to use nixgl at all anymore right? since you mentioned it as part of the steps.

Indeed, nixgl should be redundant in this configuration, as this emulates a complete mesa installation.

Regarding Void Linux, if you use OpenRC (I have done this on OpenRC under PostmarketOS), you can create an oneshot service that runs on default level and creates the symlinks.

**Comment by juipeltje** [comment]
that's what i thought, thanks again. void uses runit, but i've been messing around with void + home manager in a vm and the cron job method seems to do the trick.
